File: unicred-php/app/Domain/Cooperado/ValueObjects/Telefone.php
<?php

namespace App\Domain\Cooperado\ValueObjects;

final class Telefone
{
    public string $numero; // sempre digits-only com DDI 55


    //DDD com 2 dígitos, não iniciando com 0.

    // Móvel: 9 dígitos iniciando com 9.

    //Fixo: 8 dígitos(ex.: +55 (41) 3344-1234).

    //Bloqueia sequências repetidas (ex.: 999999999).
    public function __construct(string $numero)
    {
        $n = preg_replace('/\D+/', '', $numero) ?? '';
        if ($n === '') {
            throw new \InvalidArgumentException('Telefone inválido');
        }

        // remove DDI 55 se vier; padronizo depois
        if (str_starts_with($n, '55')) {
            $n = substr($n, 2);
        }

        // precisa ser DDD(2) + local(8|9)
        if (!preg_match('/^\d{10,11}$/', $n)) {
            throw new \InvalidArgumentException('Telefone inválido');
        }

        $ddd   = substr($n, 0, 2);
        $local = substr($n, 2);

        if ($ddd[0] === '0') {
            throw new \InvalidArgumentException('Telefone inválido');
        }

        // bloqueia locais só com o mesmo dígito
        if (preg_match('/^(\d)\1+$/', $local)) {
            throw new \InvalidArgumentException('Telefone inválido');
        }

        // regras: móvel 9 dígitos (começa com 9), fixo 8 dígitos (agora 1–9)
        if (strlen($local) === 9) {
            if (!preg_match('/^9\d{8}$/', $local)) {
                throw new \InvalidArgumentException('Telefone inválido');
            }
        } elseif (strlen($local) === 8) {
            if (!preg_match('/^[1-9]\d{7}$/', $local)) {
                throw new \InvalidArgumentException('Telefone inválido');
            }
        } else {
            throw new \InvalidArgumentException('Telefone inválido');
        }

        // padroniza para DDI 55 + DDD + local (digits-only)
        $this->numero = '55' . $ddd . $local;
    }

    public function e164(): string
    {
        return '+' . $this->numero;
    }

    public function ddd(): string
    {
        return substr($this->numero, 2, 2);
    }
}

File: java-api/src/main/java/com/unicred/cooperados/domain/vo/Telefone.java
package com.unicred.cooperados.domain.vo;

public record Telefone(String numero) {
    public Telefone {
        String n = numero==null? "": numero.replaceAll("\\D+","");
        if (n.isEmpty()) throw new IllegalArgumentException("Telefone inválido");
        if (n.startsWith("55")) n = n.substring(2);
        if (!n.matches("\\d{10,11}")) throw new IllegalArgumentException("Telefone inválido");
        String ddd = n.substring(0,2);
        String local = n.substring(2);
        if (ddd.charAt(0)=='0') throw new IllegalArgumentException("Telefone inválido");
        if (local.matches("(\\d)\\1+")) throw new IllegalArgumentException("Telefone inválido");
        if (local.length()==9) {
            if (!local.matches("^9\\d{8}$")) throw new IllegalArgumentException("Telefone inválido");
        } else if (local.length()==8) {
            if (!local.matches("^[1-9]\\d{7}$")) throw new IllegalArgumentException("Telefone inválido");
        } else throw new IllegalArgumentException("Telefone inválido");
        numero = "55"+ddd+local; //com DDI 55
    }
}
